### ISSUE TITLE
xlog/checkpoint_daemon test sometimes falls

### ISSUE THREAD

**Issue by sergw** [tarantool/tarantool#3684]
```
$ ./test-run.py checkpoint_daemon
Started ./test-run.py checkpoint_daemon
Running in parallel with 48 workers

Collecting tests in 'xlog'           (Found 1   tests): tarantool write ahead log tests.

Installing the server ...
    Found executable   at /home/s.voronezhskii/build/src/tarantool
    Found tarantoolctl at /home/s.voronezhskii/build/extra/dist/tarantoolctl

Tarantool 2.0.4-946-gef322002b
Target: Linux-x86_64-Debug
Build options: cmake . -DCMAKE_INSTALL_PREFIX=/usr/local -DENABLE_BACKTRACE=ON
Compiler: /usr/bin/cc /usr/bin/c++
C_FLAGS: -fexceptions -funwind-tables -fno-omit-frame-pointer -fno-stack-protector -fno-common -fopenmp -msse2 -std=c11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-format-truncation -fno-gnu89-inline -Werror
CXX_FLAGS: -fexceptions -funwind-tables -fno-omit-frame-pointer -fno-stack-protector -fno-common -fopenmp -msse2 -std=c++11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-format-truncation -Wno-invalid-offsetof -Werror

======================================================================================
WORKR TEST                                            PARAMS          RESULT
---------------------------------------------------------------------------------
[001] xlog/checkpoint_daemon.test.lua                                 [ fail ]
[001]
[001] Test failed! Result content mismatch:
[001] --- xlog/checkpoint_daemon.result Thu Sep 13 13:30:31 2018
[001] +++ xlog/checkpoint_daemon.reject Thu Sep 13 13:40:28 2018
[001] @@ -100,7 +100,7 @@
[001]  ...
[001]  test_run:grep_log("default", "making snapshot", 400) ~= nil
[001]  ---
[001] -- true
[001] +- false
[001]  ...
[001]  -- restore default options
[001]  box.cfg{checkpoint_interval = 3600 * 4, checkpoint_count = 4 }
[001]
[001] Last 15 lines of Tarantool Log file [Instance "xlog"][/home/s.voronezhskii/build/test/var/001_xlog/xlog.log]:
[001] 2018-09-13 13:40:28.219 [10749] main/120/console/unix/: I> set 'checkpoint_interval' configuration option to 15
[001] 2018-09-13 13:40:28.219 [10749] main/120/console/unix/: I> set 'checkpoint_count' configuration option to 20
[001] 2018-09-13 13:40:28.220 [10749] main/120/console/unix/: I> set 'checkpoint_count' configuration option to 1
[001] 2018-09-13 13:40:28.224 [10749] main/120/console/unix/: I> set 'checkpoint_count' configuration option to 2
[001] 2018-09-13 13:40:28.224 [10749] main/121/checkpoint_daemon I> reloaded
[001] 2018-09-13 13:40:28.224 [10749] main/121/checkpoint_daemon I> scheduled the next snapshot at Thu Sep 13 15:22:59 2018
[001] 2018-09-13 13:40:28.224 [10749] main/120/console/unix/: I> set 'checkpoint_interval' configuration option to 3600
[001] 2018-09-13 13:40:28.229 [10749] main/121/checkpoint_daemon I> reloaded
[001] 2018-09-13 13:40:28.229 [10749] main/121/checkpoint_daemon I> scheduled the next snapshot at Thu Sep 13 13:42:59 2018
[001] 2018-09-13 13:40:28.229 [10749] main/120/console/unix/: I> set 'checkpoint_interval' configuration option to 100
[001] 2018-09-13 13:40:28.251 [10749] main/121/checkpoint_daemon I> reloaded
[001] 2018-09-13 13:40:28.252 [10749] main/121/checkpoint_daemon I> scheduled the next snapshot at Thu Sep 13 13:59:39 2018
[001] 2018-09-13 13:40:28.252 [10749] main/120/console/unix/: I> set 'checkpoint_interval' configuration option to 1000
[001] 2018-09-13 13:40:28.258 [10749] main/121/checkpoint_daemon I> stopped
[001] 2018-09-13 13:40:28.258 [10749] main/120/console/unix/: I> set 'checkpoint_interval' configuration option to 0
[Main process] Got failed test; gently terminate all workers...
[001] Worker "001_xlog" got failed test; stopping the server...
---------------------------------------------------------------------------------
Statistics:
* fail: 1
Failed tasks:
- [xlog/checkpoint_daemon.test.lua, null]
# logfile:        /home/s.voronezhskii/build/test/var/log/001_xlog.log
# reproduce file: /home/s.voronezhskii/build/test/var/reproduce/001_xlog.list.yaml
---
- [xlog/checkpoint_daemon.test.lua, null]
...
```
After 100 runs the statistic was:
* fail: 72
* pass: 28

When I change in this code:
```
-- gh-2780 check that scheduled snapshots are performed
fiber.sleep(3 * PERIOD)
```
`3` to `8` the statistic of 100 runs was:
* pass: 100

It looks like we need a different approach of checking for state changes instead of just waiting for some timeout..

**Comment by Totktonada** [comment]
We look into that with Sergei together.

1. I think PERIOD constant should depend on workers count (test-run need to be updated to pass it via environment), because tarantool servers of different workers vie for resources and becomes less responsible. Absolute value of PERIOD is small on Linux and can be smaller then delays in parallel test run. Proposed to try PERIOD *= workers_count. Hope the test logic does not depend on absolute value of PERIOD.

2. It is undesirable to sleep for long time period, so I propose to implement one of this following ways to mitigate it:
  - save deadline time, call grep_log in a loop with small sleep and deadline time check; if a condition does not match until deadline time reached, then raise an error.
  - more general and reusable approach: make a wrapper to block until a condition meet or timeout reached, provide wrapped version of grep_log (with timeout parameter), use it accross tests.

Hope it helps.

**Comment by locker** [comment]
Fixed by 6f6a30401099aea6d8e028cb73dad1f9bb9f9c4a
